src/shapes/bsplinecurve.cpp

```diff
@@ -0,0 +1,816 @@
+            m_bbox.expand(p + r * ScalarVector3f(0, 0, 1));
+        }
+    }
+
+    std::tuple<Point3f, Vector3f, Vector3f, Vector3f, Float, Float, Float>
+    cubic_interpolation(const Float v, const UInt32 idx, Mask active) const {
+        Point4f c0 = dr::gather<Point4f>(m_control_points, idx + 0, active),
+                c1 = dr::gather<Point4f>(m_control_points, idx + 1, active),
+                c2 = dr::gather<Point4f>(m_control_points, idx + 2, active),
+                c3 = dr::gather<Point4f>(m_control_points, idx + 3, active);
+        Point3f p0 = Point3f(c0.x(), c0.y(), c0.z()),
+                p1 = Point3f(c1.x(), c1.y(), c1.z()),
+                p2 = Point3f(c2.x(), c2.y(), c2.z()),
+                p3 = Point3f(c3.x(), c3.y(), c3.z());
+        Float r0 = c0.w(),
+              r1 = c1.w(),
+              r2 = c2.w(),
+              r3 = c3.w();
+
+        Float v2 = dr::sqr(v), v3 = v2 * v;
+        Float multiplier = 1.f / 6.f;
+
+        Point3f c = (-v3 + 3.f * v2 - 3.f * v + 1.f) * p0 +
+                    (3.f * v3 - 6.f * v2 + 4.f) * p1 +
+                    (-3.f * v3 + 3.f * v2 + 3.f * v + 1.f) * p2 +
+                    v3 * p3;
+        c *= multiplier;
+
+        Vector3f dc_dv =
+            (-3.f * v2 + 6.f * v - 3.f) * p0 +
+            (9.f * v2 - 12.f * v) * p1 +
+            (-9.f * v2 + 6.f * v + 3.f) * p2 +
+            (3.f * v2) * p3;
+        dc_dv *= multiplier;
+
+        Vector3f dc_dvv =
+            (-1.f * v + 1.f) * p0 +
+            (3.f * v  - 2.f) * p1 +
+            (-3.f * v + 1.f) * p2 +
+            (1.f * v) * p3;
+
+        Vector3f dc_dvvv = -p0 + 3.f * p1 - 3.f * p2 + p3;
+
+        Float radius = (-v3 + 3.f * v2 - 3.f * v + 1.f) * r0 +
+                       (3.f * v3 - 6.f * v2 + 4.f) * r1 +
+                       (-3.f * v3 + 3.f * v2 + 3.f * v + 1.f) * r2 +
+                       v3 * r3;
+        radius *= multiplier;
+
+        Float dr_dv = (-3.f * v2 + 6.f * v - 3.f) * r0 +
+                      (9.f * v2 - 12.f * v) * r1 +
+                      (-9.f * v2 + 6.f * v + 3.f) * r2 +
+                      (3.f * v2) * r3;
+        dr_dv *= multiplier;
+
+        Float dr_dvv = (-1.f * v + 1.f) * r0 +
+                       (3.f * v - 2.f) * r1 +
+                       (-3.f * v + 1.f) * r2 +
+                       (1.f * v) * r3;
+
+        return { c, dc_dv, dc_dvv, dc_dvvv, radius, dr_dv, dr_dvv};
+    }
+
+    std::tuple<Vector3f, Vector3f, Vector3f, Vector3f>
+    partials(Point2f uv, Mask active) const {
+        Float v_global = uv.y();
+        size_t segment_count = dr::width(m_indices);
+        UInt32 segment_id = dr::floor2int<UInt32>(v_global * segment_count);
+        Float v_local = v_global * segment_count - segment_id;
+
+        Point3f C;
+        Vector3f Cv, Cvv, Cvvv;
+        Float radius, rv, rvv;
+        std::tie(C, Cv, Cvv, Cvvv, radius, rv, rvv) =
+            cubic_interpolation(v_local, segment_id, active);
+
```

Comment:
Minor: it would be great to have more information on the source/further literature regarding this calculation.